### ISSUE TITLE
Add a task instancePolicy to task runOptions.

### ISSUE THREAD

**Issue by alexr00** [microsoft/vscode#90125]
When a task has reached the max number of instances it can have (default is 1, can be changed with `instanceLimit`) the `instancePolicy` determines what happens when another instance is started.

History: https://github.com/microsoft/vscode/issues/32264
PR for `insanceLimit` which is relevant to anyone interested in implementing `instancePolicy`: https://github.com/microsoft/vscode/pull/89872

**Comment by alpalla** [comment]
I can keep working on this. Just to reiterate (basing myself off [this comment](https://github.com/Microsoft/vscode/pull/32384#issuecomment-424114086)), should `instancePolicy` provide the following options?

1. terminate the last instance
2. terminate the first instance
3. ask which instance to terminate 
4. do nothing and possibly indicate that the instance limit has been reached

I imagine the first 3 options would also have a *restart* counterpart.

**Comment by alexr00** [comment]
> 1. terminate the last instance

Yes, let's call it `terminateNewest`

> 2. terminate the first instance

`terminateOldest`

> 3. ask which instance to terminate

`prompt`? I'm not 100% on this name yet. But this should be the default and use the existing behavior with the notification.

> 4. do nothing and possibly indicate that the instance limit has been reached

I'm not sure which would be more useful here: show no indication that the instance limit has been hit or notify. Maybe we need two options: `warn` and `silent`

**Comment by alpalla** [comment]
@alexr00 I'm a little confused about the `prompt` option. 
This option asks the user which of the N active instances to terminate, correct? 
How would this option use the existing behavior with the notification?
 
So far my understanding for how to handle `prompt` would be to open the QuickPick and show all instances of the task, then when an instance is selected it would be terminated.

**Comment by alexr00** [comment]
For `prompt`, yes it would be ideal to show the quick pick instead of the notification.

**Comment by fatfatson** [comment]
waiting for improve

**Comment by darylknowles** [comment]
In the most recent release I am now getting prompted to restart or terminate when I try to run scripts from the NPM Task explorer from two different open working folders (and package.json files) that happen to share the same name.  For example, running "start" from workspace folder 1 then trying to also run "start" from workspace folder 2 causes the prompt.  I suspect we're not using the full paths to detect # of instances already running.

**Comment by aroman** [comment]
any update on this? i think there's a clear and common usecase here for running a server -- you want to terminate the previous process and restart so the port is free

**Comment by bpasero** [comment]
My primary scenario is triggering the default build task in VSCode (`Cmd+Shift+B`) a lot of times per day but currently I have to terminate the 2 running tasks from the terminal because they even run detached from the app to survive window reloads.

If there was a way to kill these 2 running tasks before starting the new ones that would be very cool 👍

**Comment by ludekvodicka** [comment]
Is this feature already implemented? Or is there any way how to force running task restart automatically when the instance is already running?

**Comment by jej** [comment]
Am I missing an option or a tweak to get rid of this pop-up?...

![capture_ecran_20220124-000525](https://user-images.githubusercontent.com/33821/150701714-7ea79920-c656-4a33-86f6-f3700a4393ec.png)

**Comment by gilescope** [comment]
Every time I click run. Surely there's a setting to just terminate the previous one when I want to run a new one without it asking every time. Every time. Like always. I mean I press run a lot. And it asks every time. Maybe we need some AI ML model to train and realize that I press the kill the existing process button after I press the run button. Did I mention it prompts every time I click run? Every time... man if I had a penny for every time it prompted me I'd be a rich man... every time. every time.

**Comment by starball5** [comment]
Related on Stack Overflow: [In VS Code, when kicking off a task which is already running, how can the task be automatically restarted?](https://stackoverflow.com/q/76125453/11107541). I'll update my answer there once this is released. Thanks, and keep up the great work!

**Comment by starball5** [comment]
@matklad nice! Note that in VS Code 1.77, you can use the builtin `runCommand` command instead of the multicommand extension (see also [this](https://stackoverflow.com/q/75808371/11107541)). I've updated [my answer to the Stack Overflow question](https://stackoverflow.com/a/76126271/11107541) to show an example based on your workaround but with my own changes.

**Comment by mattlacey** [comment]
No movement a year later? Would love to see this where I'm automatically launching an emulator after a successful build.

**Comment by lypanov** [comment]
Is there keyboard interaction via eg accessibility to access the two buttons? For me the lack of obvious keyboard interactions to do the terminate/restart is the real issue. I'd expect the dialog to take focus and the arrow keys to work. Instead I need to worsen my RSI.

Found a work around shortly after writing this:
  https://superuser.com/a/1706615
  
TL;DR: Add the snippet:

```
{
        "key": "ctrl+shift+=",
        "command": "notifications.focusToasts"
},
```

to your keybindings.json. Then ctrl-shift-b, ctrl-shift-=, tab, tab, enter.
Far from intuitive and quite laborious, but terminateAll isn't an option for me.

**Comment by BrianLeishman** [comment]
For me, (using [hugo](https://github.com/gohugoio/hugo)), the restart button doesn't even work, since it seems the port isn't free yet by the time vscode tries to start the next instance.

I always get the message "bind: address already in use". Then I have to start it again.

**Comment by barclayd** [comment]
I'm a new VSCode user having transitioned from 8 years of Webstorm usage and was shocked this still doesn't have first class support.
This is the workaround I'm using:

```json
    {
        "key": "cmd+r",
        "command": "runCommands",
        "args": {
            "commands": [
                {
                    "command": "workbench.action.tasks.terminate",
                    "args": "terminateAll"
                },
                {
                    "command": "<your_run_command_goes_here>"
                }
            ]
        }
    }
```

**Comment by dimateos** [comment]
Yep we still have to either automatically kill **all tasks** or manually click on `Terminate` or `Restart` for the specific one.

**Comment by Managor** [comment]
This would be nice to have.

**Comment by y2kbugger** [comment]
![Image](https://github.com/user-attachments/assets/56de4198-a0f9-4ec2-9747-3a59c2ec4236)

I think we all have clicked this _a lot_

**Comment by dmrickey** [comment]
I have a task to run tests for my current file. I want to kill the current task and run it with the current context (i.e. the current `relativeFile`). Not having to manually stop it before starting it again would be awesome

**Comment by aspcartman** [comment]
This is just unbelievable.

Don't the development team of vscode run tasks themselves? They have some hidden patch to get rid of this abomination that they secretly share? Do they really click w/ mouse on this toast hundreds times per day? I can't wrap my head around this, I do not understand how come this be.

+this thread has just ~100 reactions, not millions, wat is going on. Share how you work, what you do so this toast doesn't nag you? Please.

Coming from JB solutions just for Cursor's AI stuff and, well, shocked.

**Comment by iwangbowen** [comment]
What took you so long
